src/main/java/com/ait/lienzo/client/core/shape/wires/WiresConnector.java

```diff
@@ -731,6 +663,260 @@ public Point2DArray getControlPoints(){
         return m_line.getPoint2DArray();
     }
 
+    public int addControlPoint(final double x,
+                               final double y) {
+        final int index = getControlPointIndex(x, y);
+        if (index < 0) {
+            final int pointIndex = getIndexForSelectedSegment(this,
+                                                              (int) x,
+                                                              (int) y,
+                                                              getControlPoints());
+            addControlPoint(x, y, pointIndex);
+            return pointIndex;
+        }
+        return index;
+    }
+
+    public void addControlPoint(final double x,
+                               final double y,
+                               final int index) {
+        if (index > -1) {
+            final IPrimitive<?> current = getControlPoint(index);
+            if (null != current) {
+                final Point2D currentLocation = current.getLocation();
+                if (currentLocation.getX() == x &&
+                        currentLocation.getY() == y) {
+                    return;
+                }
+            }
+            final Point2DArray oldPoints = getControlPoints();
+            final Point2DArray newPoints = new Point2DArray();
+            for (int i = 0; i < oldPoints.size(); i++) {
+                if (i == index) {
+                    newPoints.push(new Point2D(x, y));
+                }
+                newPoints.push(oldPoints.get(i));
+            }
+            m_line.setPoint2DArray(newPoints);
+        }
+    }
+
+    public void destroyControlPoints(final int[] indexes) {
+        final Point2DArray oldPoints = getLine().getPoint2DArray();
+        final Point2DArray newPoints = new Point2DArray();
+        for (int i = 0; i < oldPoints.size(); i++) {
+            if (!contains(indexes, i)) {
+                newPoints.push(oldPoints.get(i));
+            }
+        }
+        getLine().setPoint2DArray(newPoints);
+    }
+
+    private static boolean contains(final int[] indexes,
+                                    final int index) {
+        for (int i : indexes) {
+            if (i == index) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public void moveControlPoint(final int index,
+                                 final Point2D location) {
+        getControlPoints().get(index).set(location);
+        final IPrimitive<?> point = getControlPoint(index);
+        if (null != point) {
+            point.setLocation(location);
+        }
+    }
+
+    private IPrimitive<?> getControlPoint(final int index) {
+        return getControlPoint(index, null);
+    }
+
+    private IPrimitive<?> getControlPoint(final Integer index,
+                                          final Point2D location) {
```

Comment:
location parameter of this private method is always `null` but inside of method I can see `location.getX()`, `location.getY()`. So there is dead code with useless parameter OR NullPointer exceptions.

Reply:
well yes and no.. :) I mean it does two checks: first check if the index is present, otherwise it assumes the location will be never null and also it's a private method, so I can guarantee that this NPE will never happens from the class itself, so why the need for another NPE check? Are just making the code look less clear to read...
Anyway minor thing, just giving my opinion, I can change thsi np, but WDYT?

Reply:
I guess it should be like this:

```java
    private IPrimitive<?> getControlPoint(final int index) {
        return getPointHandles().getHandle(index).getControl();
    }

    private IPrimitive<?> getControlPoint(final Point2D location) {
        for (IControlHandle handle : getPointHandles()) {
            if (handle.getControl().getLocation().equals(location)) {
                return handle.getControl();
            }
        }

        return null;
    }
```

But, actually  getControlPoint#(Point2D) is never used in the code, so it can be even simpler:
```java
    private IPrimitive<?> getControlPoint(final int index) {
        return getPointHandles().getHandle(index).getControl();
    }
```